crates/huub/src/solver/engine.rs

```diff
@@ -324,20 +324,22 @@ impl PropagatorExtension for Engine {
 		}
 
 		// Run propgagators to find any conflicts
-		ctx.run_propagators(&mut self.propagators);
-		// No propagation can be triggered (all variables are fixed, so only
-		// conflicts are possible)
-		debug_assert!(self.state.propagation_queue.is_empty());
-
-		// Process propagation results, and accept model if no conflict is detected
-		let accept = if let Some(conflict) = self.state.conflict.clone() {
-			// Move conflict to clauses before backtrack
-			self.state.clauses.push_back(conflict);
-			false
-		} else {
-			true
-		};
-
+		let mut accept = true;
+		loop {
+			ctx.run_propagators(&mut self.propagators);
+			// No propagation can be triggered (all variables are fixed, so only
+			// conflicts are possible)
+			debug_assert!(ctx.state.propagation_queue.is_empty());
+			// Process propagation results, and accept model if no conflict is
+			// detected
+			if let Some(conflict) = ctx.state.conflict.take() {
+				// Move conflict to clauses before backtrack
+				ctx.state.clauses.push_back(conflict);
```

Comment:
The normal clauses in `state.clauses` and conflict clauses in `state.conflict` seems to be handled differently in `add_external_clause`. Shall we change the `conflict` variable into a vector to ensure the consistent handling?

Reply:
The difference is that `state.clauses` are persistent, and will be kept even when you backtrack. We could change `conflict` to be a vector and capture all conflicts immediately, but it would still require the move to `state.clauses` because we immediately backtrack afterwards in `check_solutions`. (Note that the other difference is that all other clauses and propagations are posted before the conflict, to ensure it will cause conflict in the SAT solver. However, since conflicts are added to the end of the list and there cannot be any propagations, this is not a problem).

So if we stick to my proposed design, where we detect all conflicts in `check_solutions`, but find only the first in `propagate`, then don't think we need to change it.

Reply:
I see. Probably the performance change is due to the search heuristic. Let me test it using a larger benchmark and also try the ``--vsids-only`` option.

Reply:
That would be great! Thanks :)

Reply:
Tested with jobshop and openshop instances. The performance seems to remain the same.

Reply:
If the performance is just the same, then I'm not sure whether the change is worthwhile.

Like we discussed in person: only one of the clauses is used in conflict resolution, so any other given clauses are added to the clause database “as is”. I'm not sure whether this wouldn't be something we would hope to avoid.

I'll at least make another branch that ensures that conflict clauses in `check_solution` are still added as “removable” clauses.